Add gate to disable control token decoding

ju6ge opened the pull request:
This PR implements the required changes to address #826

For easy opt in to the original behavior I added a check to an environment variable. This would make the behavior controllable 
at runtime. I can easily change this to be a compile time opt in by using a rust feature flag. Let me know what you think …

Kind regards
ju6ge

ju6ge committed:
use environment variable to disable control token decoding

llama-cpp-rs original usage required ommiting control tokens from the
consumer of the library. This should not be the default though so now
this behavior can be selectively be enabled through an environment variable

MarcusDunn reviewed (changes requested):
Thanks for the PR. See comment.

ju6ge committed:
instead of an env variable use `special` parameter

given that the `special` function argument is used to toggle if the
cpp bindings to llama.cpp render special tokens to the output the flag
can also be reused to feature gate the exclusion of `token_bos` and
`token_eos` from the output.

ju6ge committed:
introduce new variant to `Special` enum

to keep all ranges of behavior possible a special variant to the
`Special` enum was introduced (ExcludeBosAndEos). It allows decoding of tokens but excludes
`bos` and `eos` tokens from the stream. This can be used to keep the old
behavior of llama-cpp-rs for decoding streams while allowing the
expected behavior that all tokens will be decoded by default.

See https://github.com/utilityai/llama-cpp-rs/pull/856 for the
discussion about this.

MarcusDunn reviewed (changes requested):
I think we should deprecate the current function (as well as all of our functions that call it), create a new versions that call `token_to_peice` without any special logic. (no messing around with attrs, as I imagine this is pretty application specific)

This aligns better with the goal of being a safe wrapper and once the deprecated functions are removed there is less code to maintain.

`Special` should remain an enum of two variants.
